Feature/spotlight 2.2.1

jhriv opened the pull request:
Fixes #23, #30

Completely revamps the way rails apps are deployed. 

1. The app itself is a completely separate repository. No more trying to roll it in the playbook.
2. Proper systemd service files, for consistent starting at boot, and in case of failure.
3. Local testing supports multiple OSes simultaneously

Please review: (no. seriously. take a look. run it. read the docs. if there's a question **that is a bug**)
@ucsdlib/developers
@ucsdlib/operations

jhriv committed:
Upgrade to Vagrant as Infrastructure v0.4.0

jhriv committed:
Move vars to group_vars

Prevents having to add vars_files: to each playbook
Makes is easier to override on a per-host basis

jhriv committed:
Add deploy playbook and hooks

jhriv committed:
Add rbenv playbook

jhriv committed:
Add nginx

jhriv committed:
Add postgresql server

Only required for Vagrant systems

jhriv committed:
Rename setup to hostname

Limit scope of setup to only change the hostname
Only required for Vagrant

jhriv committed:
Remove deprecated commands playbook

jhriv committed:
Start puma service, outside of deploy

jhriv committed:
Factor solr into separate playbook

jhriv committed:
Remove deprecated install playbook

jhriv committed:
Factor out vars_files to group_vars

jhriv committed:
Factor out OS packages to separate playbook

jhriv committed:
Add required roles

Outstanding bug: ucsdlib.solr needs to be updated to use OS java
instead of using Oracle java

jhriv committed:
Ignore vault and roles

jhriv committed:
Remove postgress database, and re-add it

For debugging only

jhriv committed:
Removed debugging playbook

It's still in history, if you really want to grab it again.

jhriv committed:
Add a better postgres-reset.yml, one that works

The previous incarnation tried to assign the deleted database
to the user. We override the user for the database deletion to avoid
such a calamtiy.

jhriv committed:
Remove debugging postgres-reset.yml playbook

Use the previous commit's version - it's better

$ git checkout 4f1b840725c2bb2b371a1c5b463e8548ad6d7fc5 -- postgres-reset.yml

This is assuming the previous commit's hash is the above.

jhriv committed:
Move vagrant files to separate directory

Keeping the top level directory a little cleaner, and maintaining
a brighter line between app and vagrant-sepcific requirements

jhriv committed:
Move vagrant specifit playbooks to vagrant area

jhriv committed:
Fix user gem installation

rbenv and ansible's gem module do not get along
Using the rbenv gem shim works

Two methods are used here, but only one needs to be. I lead towards
using the gem shim. Using default-gems seems to ignore changes unless a
new ruby is built.

jhriv committed:
Update vars.yml to support gem dependency format

Explcitily list gems and their versions.
Versions are optional

jhriv committed:
Bundle gems / create solr core as part of deploy

jhriv committed:
Remove auto-generated default-gems

jhriv committed:
Elevate privileges to install OS packages

jhriv committed:
Move solr core creation to ansistrano.deploy

jhriv committed:
Add master(site) playbooks

jhriv committed:
Move vagrant specific ignores to vagrant

jhriv committed:
Update documentation

The documentation should match the code. This is at least closer.

jhriv committed:
Rename site to main

Site was always a bad name. Main is far more standard, and expected.

jhriv committed:
Add RHEL(CentOS) support

jhriv committed:
Fix typo

jhriv committed:
Move vagrant setup helper to vagrant directory

jhriv committed:
Move ansistrano hoks to separate directory

All about keeping the top level neat and tidy

jhriv committed:
Add RAILS_MASTER_KEY

Puma didn't need it, but rils console did

jhriv committed:
Rename systemd service file to match application

Supports having multiple apps on a single host, all with
separate and individual puma daemons

jhriv committed:
Replace SECRET_KEY_BASE with RAILS_MASTER_KEY

credentials.yml.enc contains the value for secret_key_base
Rails 5.1 introduced encrypted secrets, which rails new creates

jhriv committed:
Update to vagrant-as-infrastructure v0.4.1

Works around a new Vagrant bug

jhriv committed:
Update ucsdlib.solr role

Removes hard java role dependency

jhriv committed:
Enable full CentOS support

Update vagrant-as-infrastructure 0.6.0 (ansible groups)
Combine ubuntu/centos into one Vagrantfile
Fix CentOS $HOME permissions / SELinux to permissive
Add CentOS gemfile build dep (G++)
Work around CentOS NGINX config depclaring it's own default server
Add NodeSource repository (nodejs on CentOS)

jhriv committed:
Bump ucsdlib.solr version (fixes local sudo bug)

jhriv committed:
Remove debugging code

mcritchlow commented on README.md:
```diff
@@ -15,51 +15,52 @@ Attempting to get a working production prototype of [Project Blacklight][BL] + [
 ## Playbooks
 
-* `install.yml` Installs and starts Solr, sets up the Blacklight rails Application
+* `after-*-sym.yml` Ansistrano.deploy hooks
```
For those unfamiliar, maybe link here, or elsewhere, to Ansistrano itself? As far as I can tell this is where it is introduced. This could be https://github.com/ansistrano/deploy#hooks-custom-tasks perhaps?

jhriv replied:
Fixed.

mcritchlow commented on deploy-hooks/after-current-sym.yml:
```diff
@@ -0,0 +1,23 @@
+---
+
+- name: Install systemd unit file
```
Should the name call out puma explicitly as the systemd service being installed?

jhriv replied:
I thought about that, and originally used puma. The two problems are 
1) we don't care about what app server is used, only the end service (in this case, spotlight) 
2) there is no process named puma anyway. the process is bundle.

three (bonus problem!) - if there are multiple apps, they each need an app server. puma, puma-1, puma-2 is confusing. Which puma? puma-spotlight, puma-dams is clear, but extraneous information.

mcritchlow replied:
Yeah fair enough

mcritchlow commented on deploy-hooks/after-shared-sym.yml:
```diff
@@ -0,0 +1,55 @@
+  args:
+    chdir: '{{ ansistrano_release_path.stdout }}'
+  changed_when: true
+  when: '( app_db_migration_status.stdout_lines[-1] | regex_search("^  down ")) != none'
```
I like this very much :+1:

mcritchlow commented on group_vars/all/vars.yml:
```diff
@@ -0,0 +1,92 @@
+# ansitrano+git
+ansistrano_deploy_via: 'git'
+ansistrano_git_repo: 'https://github.com/ucsdlib/spotlight-exhibits-demo.git'
+ansistrano_git_branch: 'develop' # What version of the repository to check out. This can be the full 40-character SHA-1 hash, the literal string HEAD, a branch name, or a tag name
```
We should probably make this `master` at some point and drop `develop`. We've recently switched to using a `master` mainline branch + releases as tags.

jhriv replied:
It's in a vars file, so a more master deploy can use `ansible-playbook -e ansistrano_git_branch=master` or even tag.

Can also be in an host_vars/<name-of-prod-server>/spotlight.yml.

mcritchlow replied:
I guess all I'm saying is that for other development repos we don't have `develop` branches at all anymore. This repo doesn't necessarily have to have that convention. Just noting the distinction.

jhriv replied:
Have we abandoned git-flow entirely?

mcritchlow commented on ruby.yml:
```diff
@@ -0,0 +1,37 @@
+
+  roles:
+    - { name: zzet.rbenv, 
+        rbenv_users: [ vagrant ],
```
Should the `vagrant` string here be `ansible_user_id`?

jhriv replied:
Did that not get fixed? I think I fixed that.

Pretty sure since I found that error when deploying to vSphere

jhriv replied:
I did fix it, Hadn't pushed it yet.

jhriv replied:
NowI have !

mcritchlow commented on vagrant/main.yml:
```diff
@@ -0,0 +1,11 @@
+- import_playbook: hostname.yml
+- import_playbook: centos.yml
+- import_playbook: postgres.yml
+  #- import_playbook: playbooks/solr.yml
```
do we need these commented out entries or could we remove them?

jhriv replied:
They can be removed.

mcritchlow commented on vagrant/README.md:
```diff
@@ -0,0 +1,15 @@
+Spotlight for Exhibits Demo on Vagrant
```
I'm just generally curious what your thoughts are as to the role of this `README` as opposed to the project root version. Do we need both? Should this one hold the more vagrant specific things ( so more easily reused between projects for example ) and the root level be more Spotlight specific stuff? Just kinda thinking out loud.

jhriv replied:
Right now it's part of spotlight. When we apply it to something else (Hyrax, looking at you) we can learn where the specifics are, and abstract them away, and make something more generic. Perhaps even as a sub-module (even though I don't really like sub-modules)

mcritchlow replied:
Sounds good to me

mcritchlow reviewed (commented):
This looks excellent. Some general questions, but :+1:

jhriv committed:
Abstract out inadvertent hardcoded names

Or, "Lessons Learned from Assuming All The World's a Vagrant"

Revamp the nginx config
Based off of "Deploy Rails 5.2 to Amazon EC2 with Puma, Nginx and
Capistrano 3 for multiple environments"
https://colindao.wordpress.com/2018/03/30/tu-dong-deploy-rails-app-voi-capistrano-unicorn-va-nginx/

jhriv committed:
Remove stale, commented-out playbook imports

jhriv committed:
Add link to Ansistrano hook documentation

jhriv committed:
Deploy from master branch

Every day we stray further from git-flow's light.

mcritchlow reviewed (approved):
:100:
